### ISSUE TITLE
Allow re-rendering Markdown preview when updated extra configurations provided by extension

### ISSUE THREAD

**Issue by yhatt** [Microsoft/vscode#70936]
If Markdown extension provides some extra settings to affect the rendering of Markdown, it would want to re-render Markdown preview when changed configuration values. In fact, VS Code built-in configurations (e.g. `markdown.preview.breaks`) will behave like that. (Tested in 1.33.0-insider)

![breaks](https://user-images.githubusercontent.com/3993388/54802572-41e49d00-4caf-11e9-9c9c-41aff19c148a.gif)

I have met with such a case while creating [a Markdown extension that has configurable rendering](https://github.com/marp-team/marp-vscode), but I could not find a way to re-trigger Markdown rendering manually. `markdown.preview.refresh` command is meaningless because it just refresh WebView and not triggered a rendering process.

```typescript
import { ExtensionContext, commands, workspace } from 'vscode'

export const activate = ({ subscriptions }: ExtensionContext) => {
  subscriptions.push(
    workspace.onDidChangeConfiguration(e => {
      if (e.affectsConfiguration('extension.configuration')) {
        // FIXME: `markdown.preview.refresh` won't re-trigger rendering
        commands.executeCommand('markdown.preview.refresh')
      }
    })
  )

  return {
    extendMarkdownIt: md => {
      md.use(() => {
        const conf = workspace.getConfiguration('extension').get('configuration')
        // ... and use configured value in markdown-it plugin ...
      })
    }
  }
}
```

Thus, user have to change Markdown buffer once for previewing with updated configurations. Probably many users would expect to reflect the changed configuration to preview immediately.

![](https://user-images.githubusercontent.com/3993388/54802890-9a686a00-4cb0-11e9-8a76-e84403a051bc.gif)

It would be awesome if there was a command (or API for extension) to re-trigger rendering of Markdown.

**Comment by mjbvz** [comment]
PRs welcome to make the refresh command re-render

**Comment by vitaliymaz** [comment]
Reproducible for me, can I work on the PR?

**Comment by vitaliymaz** [comment]
The problem happens with cached tokens here 
https://github.com/Microsoft/vscode/blob/2f2085f3e1e3c75916ea950d5c8c2e1e4c28b3dd/extensions/markdown-language-features/src/markdownEngine.ts#L121

It makes a decision whether to recalculate tokens based on the document version and markdown extension configurations. But it doesn't consider changing configurations of markdownItPlugins.

There are two possible solutions I see:
1. Extract configurations of all plugins which extend markdown base plugin and recalculate tokens if something changed there.
2. Clean tokens cache on refresh command or pass some force flag. 

@mjbvz , maybe you can help me with a decision.

**Comment by yhatt** [comment]
IMO I prefer 2, a controllable way in per extension.

Some settings defined in Markdown extension would not require clearing cache. I'm afraid to degraded performance.

In addition, I have an extension that must trigger re-rendering Markdown by VSCode native configuration `window.zoomLevel`. It does not defined in the extension. (https://github.com/marp-team/marp-vscode/issues/8#issuecomment-474656886)

**Comment by mjbvz** [comment]
@vitaliymaz I'm in favor of the second approach as well: making the `refresh` command flush the cache. That way it's entirely up to the extension to say when the preview is refreshed

**Comment by vitaliymaz** [comment]
@mjbvz, thank you for the feedback. Working on the second approach.

**Comment by mjbvz** [comment]
Fixed by #71442

**Comment by yhatt** [comment]
Thanks @mjbvz, and @vitaliymaz's awesome work! I'll start to test in the next insider version. Cheers :beer:
